### ISSUE TITLE
Kubernetes platform support

### ISSUE THREAD

**Issue by gitdhar** [jpmorganchase/daytrader-example-quotesrepo#1]
Is there an updated description of how to deploy this on Kubernetes?

**Comment by dhvines** [comment]
README.MD files are in the repo.

Sent from my iPhone

On Mar 22, 2020, at 9:28 PM, gitdhar <notifications@github.com<mailto:notifications@github.com>> wrote:


Is there an updated description of how to deploy this on Kubernetes?

**Comment by gitdhar** [comment]
We have Kube cluster using Virtualbox VMs (one master and 2 workers) on a server that is behind a firewall. 
In the pom.xml, I see that we specify "localhost". 
I need to have the service (that is hosted on 127.0.0.1:8001) be accessed by an external IP. 
What are the changes I need to access this service on an external IP?

I modified the ./env/external/kubernetes/conf/service.yaml to include an externalIP like so:
"
  ports:
  - port: 443
    protocol: TCP
    targetPort: 4443
  externalIPs:
  - 192.168.205.10"
This exposed the quotes service. However, this is only the quotes service and not the App (that is on 127.0.0.1:8001)

**Comment by dhvines** [comment]
Two comments:


  1.  For accessing (and functional testing) the application external to the cluster, an ingress resource is used. See https://github.com/jpmorganchase/daytrader-example-webrepo/blob/master/README.MD. That ingres resources gets you from outside the cluster to the web application and from there, the web application uses cluster IP services to communicate with the rest of the microservices.
  2.  For integration testing purposes, the maven pom file, uses the maven failsafe plugin and shell scripts to start the kubectl proxy prior to running the integration tests. All external requests (during the integration test) then flow to the api server via the local proxy (which by default is listening on localhost:8001) and it is the api server that forwards traffic to the Cluster IP services. If you need a different port (than 8001) then you can change the aforementioned shell script to start the kubectl proxy to start listening on a different port (see the kubectl proxy CLI for more details). Or if you are having trouble connecting to the API server, then you can change the startup script from usng kubectl proxy to using kubectl port-forwarding (again see the API documentation for additionale details on kubectl port-forward). If you make this latter change then you will also need to change the URLs (only in  the POM) to use the Cluster IP FQDN.

Don

________________________________

**Comment by dhvines** [comment]
The web microservices already has an ingress resource. It exposes the web outside the cluster. You can add another path to that ingress resource to expose quotes or you can create a load balancer service just for quotes... personally I’d go with the ingress controller...

Sent from my iPhone

On Mar 24, 2020, at 2:31 AM, gitdhar <notifications@github.com<mailto:notifications@github.com>> wrote:


We have Kube cluster using Virtualbox VMs (one master and 2 workers) on a server that is behind a firewall.
In the pom.xml, I see that we specify "localhost".
I need to have the service (that is hosted on 127.0.0.1:8001) be accessed by an external IP.
What are the changes I need to access this service on an external IP?

I modified the ./env/external/kubernetes/conf/service.yaml to include an externalIP like so:
"
ports:

  *   port: 443
protocol: TCP
targetPort: 4443
externalIPs:
  *   192.168.205.10"
This exposed the quotes service. However, this is only the quotes service and not the App (that is on 127.0.0.1:8001)

**Comment by gitdhar** [comment]
Thanks. We went with the ingress resource for the example-webrepo.
> [root@tester2 k8s-cluster]# kubectl get services
> NAME            TYPE        CLUSTER-IP      EXTERNAL-IP      PORT(S)   AGE
> daytrader-web   ClusterIP   10.106.234.11   192.168.205.10   443/TCP   49s

I can now access the service from outside using the https:192.168.205.10:443
However, when I try to Login or Register a new user, it fails,

> "TradeServletAction.doLogin(...)Exception logging in user=uid:0 with password=xxx; java.net.UnknownHostException: daytrader-gateway"
> 

What am I missing here?

**Comment by dhvines** [comment]
You have to populate the database. Follow the instructions in the readme for running the functional test in the web readme.md file.

Sent from my iPhone

On Mar 24, 2020, at 9:23 PM, gitdhar <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks. We went with the ingress resource for the example-webrepo.

[root@tester2 k8s-cluster]# kubectl get services
NAME TYPE CLUSTER-IP EXTERNAL-IP PORT(S) AGE
daytrader-web ClusterIP 10.106.234.11 192.168.205.10 443/TCP 49s

I can now access the service from outside using the https:192.168.205.10:443
However, when I try to Login or Register a new user, it fails,

"TradeServletAction.doLogin(...)Exception logging in user=uid:0 with password=xxx; java.net.UnknownHostException: daytrader-gateway"

What am I missing here?

**Comment by dhvines** [comment]
The app was originally used as an industry benchmark so it is used by creating and populating the  dB then running the load test. So the functional test isnsimilar you crest and populate the dB the run the fn test...

Sent from my iPhone

On Mar 24, 2020, at 9:23 PM, gitdhar <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks. We went with the ingress resource for the example-webrepo.

[root@tester2 k8s-cluster]# kubectl get services
NAME TYPE CLUSTER-IP EXTERNAL-IP PORT(S) AGE
daytrader-web ClusterIP 10.106.234.11 192.168.205.10 443/TCP 49s

I can now access the service from outside using the https:192.168.205.10:443
However, when I try to Login or Register a new user, it fails,

"TradeServletAction.doLogin(...)Exception logging in user=uid:0 with password=xxx; java.net.UnknownHostException: daytrader-gateway"

What am I missing here?

**Comment by gitdhar** [comment]
Do we hae to deploy each of these repo's separately? 
`daytrader-example-webrepo, daytrader-example-quotesrepo, daytrader-example-accountsrepo, daytrader-example-portfoliorepo `
After this, we need to populate the database, correct?

**Comment by dhvines** [comment]
Yes. They are independent microservices and as such have their own separate lifecycle. I was going to write a helm umbrella chart so you have the option to deploy them independently or to deploy them as a whole.

Don

________________________________

**Comment by gitdhar** [comment]
I'm sorry (since we're not too familiar with this app).
Please let us know, how we can deploy all these services so that we have the complete app running on our Kubernetes env (running on a Vagrant cluster).

**Comment by dhvines** [comment]
Follow the README.MD files in each of the repos ... The startup order is:


  1.  Quotes
  2.  Portfolios
  3.  Accounts
  4.  Gateway
  5.  Web ...

If you stat with the WEB README.MD it tells you the startup order in the run functional test section.

Don

________________________________

**Comment by dhvines** [comment]
Yes. That is right but it installs them individually ... a helm umbrella chart would deploy them all ...

Don

________________________________

**Comment by gitdhar** [comment]
Thanks; going through one-by-one service install in the order, we are now able to run the entire app!
Can you please give us some info about how to generate a sample perf load that exercises the different services in this app?

**Comment by dhvines** [comment]
You can find the original daytrader jmeter scripts in the following link. You will need to make a few minor changes to them since the original daytrader monolith was stateful, but the microservices are stateless. In essence you will need to pass a couple of cookies in the jmeter web request. I did that internally at JPM but was not able to release them to the public domain ...

https://developer.ibm.com/wasdev/docs/measuring-performance-daytrader-3-benchmark-sample/

Measuring performance with the Daytrader 3 benchmark sample - WASdev<https://developer.ibm.com/wasdev/docs/measuring-performance-daytrader-3-benchmark-sample/>
The Daytrader 3 benchmark sample provides a suite of test cases to measure scalability and performance. This article explains how to run a web-based workload driver (Apache JMeter) against it to measure throughput performance.
developer.ibm.com


________________________________

**Comment by gitdhar** [comment]
Can we set this up to run the Accounts, Portfolio, and Quotes services un-encrypted?

**Comment by gitdhar** [comment]
@dhvines , we're running into some issues in setting this up to run on http (not https). The webapp is open on http, but fails to communicate with the database. Can you please list the things we need to do to export services on http instead?

**Comment by dhvines** [comment]
In the daytrader-quotes/src/main/resources/application.properties you will need to set server.ssl.enabled to false ...

If you're deploying to kubernetes you will need to change the scheme in the daytrader-quotes/env/external/kubernetes/conf/config.yaml

If you running it on your desktop (spring boot) then you may need to change startup scripts the daytrader-quotes/env/external/springboot/bin folder

I'd suggest getting the springboot app working first on your local desktop then getting it to work on kubernetes one

Don

________________________________

**Comment by gitdhar** [comment]
With all this, only the `daytrader-web` works but not others. When we try to recreate a table it hangs and times out eventually. 

We're running on Kubernetes. 
There is no file - `daytrader-quotes/env/external/kubernetes/conf/config.yaml`. Did you mean `daytrader-example-quotesrepo/daytrader-quotesapp/daytrader-quotes/env/external/kubernetes/conf/deployment.yaml`

We made changes to `application.properties` for each service. 
Use the same 443 port but with http instead of https (hope that is not a problem).

```
root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-gateway:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-accounts:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-portfolios:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-quotes:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-web:443/health
curl: (7) Failed to connect to daytrader-web port 443: Connection timed out
```

Looked at kubectl logs for each of these pods. 
Logs on web show
`03:07:19.524[26]: GatewayRemoteCallService.recreateDBTables() - http://daytrader-gateway/admin/recreateDBTables`
But kubectl logs in gateway (and all others) shows nothing after the init.
```
2020-04-12 22:28:48.768  INFO 1 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 2443 (http)
2020-04-12 22:28:48.774  INFO 1 --- [           main] o.a.g.d.j.gateway.GatewayApplication     : Started GatewayApplication in 12.124 seconds (JVM running for 13.507)
2020-04-12 22:29:06.153  INFO 1 --- [nio-2443-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring FrameworkServlet 'dispatcherServlet'
2020-04-12 22:29:06.154  INFO 1 --- [nio-2443-exec-1] o.s.web.servlet.DispatcherServlet        : FrameworkServlet 'dispatcherServlet': initialization started
2020-04-12 22:29:06.986  INFO 1 --- [nio-2443-exec-1] o.s.web.servlet.DispatcherServlet        : FrameworkServlet 'dispatcherServlet': initialization completed in 832
```
Any ideas why the connection between web and gateway seems to be broken? Note that the healthcheck probe from the web to gateway and others seems fine. Only the self healthcheck on web is not working.

```
root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-gateway:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-accounts:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-portfolios:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-quotes:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-web:443/health
curl: (7) Failed to connect to daytrader-web port 443: Connection timed out
```

**Comment by dhvines** [comment]
Thanks for the message. Will look into this first thing on wed morning


Sent from my iPhone

On Apr 13, 2020, at 11:14 PM, gitdhar <notifications@github.com> wrote:

﻿

With all this, only the daytrader-web works but not others. When we try to recreate a table it hangs and times out eventually.

We're running on Kubernetes.
There is no file - daytrader-quotes/env/external/kubernetes/conf/config.yaml. Did you mean daytrader-example-quotesrepo/daytrader-quotesapp/daytrader-quotes/env/external/kubernetes/conf/deployment.yaml

We made changes to application.properties for each service.
Use the same 443 port but with http instead of https (hope that is not a problem).

root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-gateway:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-accounts:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-portfolios:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-quotes:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-web:443/health
curl: (7) Failed to connect to daytrader-web port 443: Connection timed out


Looked at kubectl logs for each of these pods.
Logs on web show
03:07:19.524[26]: GatewayRemoteCallService.recreateDBTables() - http://daytrader-gateway/admin/recreateDBTables
But kubectl logs in gateway (and all others) shows nothing after the init.

2020-04-12 22:28:48.768  INFO 1 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 2443 (http)
2020-04-12 22:28:48.774  INFO 1 --- [           main] o.a.g.d.j.gateway.GatewayApplication     : Started GatewayApplication in 12.124 seconds (JVM running for 13.507)
2020-04-12 22:29:06.153  INFO 1 --- [nio-2443-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring FrameworkServlet 'dispatcherServlet'
2020-04-12 22:29:06.154  INFO 1 --- [nio-2443-exec-1] o.s.web.servlet.DispatcherServlet        : FrameworkServlet 'dispatcherServlet': initialization started
2020-04-12 22:29:06.986  INFO 1 --- [nio-2443-exec-1] o.s.web.servlet.DispatcherServlet        : FrameworkServlet 'dispatcherServlet': initialization completed in 832


Any ideas why the connection between web and gateway seems to be broken? Note that the healthcheck probe from the web to gateway and others seems fine. Only the self healthcheck on web is not working.

root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-gateway:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-accounts:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-portfolios:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-quotes:443/health
{"status":"UP"}root@daytrader-web-db8d999f6-494kx:/# curl -k http://daytrader-web:443/health
curl: (7) Failed to connect to daytrader-web port 443: Connection timed out

**Comment by dhvines** [comment]
Sorry, other things came up. Will  get back this afternoon..



________________________________

**Comment by dhvines** [comment]
Ok.

Made changes to run quotes with http only (not https).Here are the changes:


  1.  Global search and replace in the daytrader-quotes project for *.* and replace https with http (make sure to match case)
  2.  Regenerate the jks files. To do go to the daytrader-quotes/env/external/ssl folder and run "create_keystores.sh [.bat]
  3.  Copy and replace *.jks from daytrader-quotes/env/external/ssl to daytrader-quotes/src/main/resources/external/ssl
  4.  Open daytrader-quotes/src/main/resources/application.properties for editing
  5.  Change sserver.ssl.enabled=true to server.ssl.enabled=false
  6.  Run mvn clean install from the daytrader-quotes*app* folder
  7.  You should see

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for daytrader-quotesapp 4.0.0:
[INFO]
[INFO] daytrader-quotesapp ................................ SUCCESS [  0.342 s]
[INFO] daytrader-utilities ................................ SUCCESS [  2.132 s]
[INFO] daytrader-entities ................................. SUCCESS [  0.657 s]
[INFO] daytrader-quotes ................................... SUCCESS [ 42.503 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  46.081 s
[INFO] Finished at: 2020-04-16T12:40:20-05:00
[INFO] ------------------------------------------------------------------------


After getting the springboot build to work with http then let's talk about kubernetes ok?

Thanks,

Don


________________________________

**Comment by dhvines** [comment]
Please make sure you see this in the output from the build.

    Results :

    Tests run: 5, Failures: 0, Errors: 0, Skipped: 0

Note that it'll be above the build outputs shown in earlier email.

That indicates that all 5 integration tests were successful.

________________________________

**Comment by dhvines** [comment]
Hmm. I pulled down latest copy and discovered a couple of other changes. These must have been cut and past errors. Anyway, they are:


  1.  Open daytrader-quotes/env/external/springboot/bin/setenv.sh[.bat] for editing
  2.  Change export SERVER_PORT=3443 to set SERVER_PORT=4443 and
  3.  Change export SERVER_PORT_HTTPS=3443 to set SERVER_PORT_HTTPS=4443 and
  4.  Change export DAYTRADER_APP_ARTIFACTID=daytrader-portfoliosapp to export DAYTRADER_APP_ARTIFACTID=daytrader-quotesapp and
  5.  Change export DAYTRADER_WAR_ARTIFACTID=daytrader-portfolios to export DAYTRADER_WAR_ARTIFACTID=daytrader-quotes

Sorry for these changes. But that should get you running in sprngboot with http ... After that, I'll help you get it running on kubernetes ...

Don
________________________________

**Comment by gitdhar** [comment]
Don,
We’ve tried to modify the deployment and service yamls to use http scheme but without changing the port from 443. That didn’t work. However when we change the ports to use 80 instead of http it all worked smoothly (on Kubernetes). 
We really needed this on Kube and hence the effort.
Thanks

**Comment by dhvines** [comment]
No problem. Glad you got it working.

________________________________

**Comment by gitdhar** [comment]
@dhvines , need some help with the jmeter scripts (daytrader3.jmx from https://github.com/WASdev/sample.daytrader3 as per your advice above) that we need to run against this app on Kubernetes. 
I tried using the daytrader3.jmx but it doesnt work. (Err 100%). Can you please let us know the changes we need in the daytrader3.jmx

**Comment by gitdhar** [comment]
@dhvines , we were able to setup Jmeter and record a typical user scenario. However, I need some help in creating an "atypical" scenario. Can you please suggest some ways of doing this? Would be happy to provide more detail over email too. Thanks

**Comment by dhvines** [comment]
Yes. I was abgle to get the orignal jmeter script working.

However, I changed the web UI from using session state to using cookies.

So you will need to change the load scripts from uses sessions to using cookies.

Company did not allow me to release those load tests back into the open source project.

Perhaps you can do so when you get them working?

Don

________________________________

**Comment by gitdhar** [comment]
We recorded a few typical interactions using Jmeter and are replaying it. Did not have to do any special handling for Sessions/Cookies. How do you suggest I verify that things are working as expected with Jmeter?
Also, referring back to my question earlier, we need to create some malicious user interactions (exfiltrating data, etc.). Can you please suggest any ideas?

**Comment by dhvines** [comment]
The JMeter scripts akready exist. You can get them

 https://github.com/WASdev/sample.daytrader7/tree/master/jmeter_files

Start with those and modify the management of user connections In JMeter from using sessions to using cookies.

There is also a lot of documentation available for Inline for how to run them. See this link for one such example:

https://developer.ibm.com/wasdev/docs/measuring-performance-daytrader-3-benchmark-sample/



Sent from my iPhone

On Apr 27, 2020, at 10:58 AM, gitdhar <notifications@github.com> wrote:

﻿

We recorded a few typical interactions using Jmeter and are replaying it. Did not have to do any special handling for Sessions/Cookies. How do you suggest I verify that things are working as expected with Jmeter?
Also, referring back to my question earlier, we need to create some malicious user interactions (exfiltrating data, etc.). Can you please suggest any ideas?

**Comment by gitdhar** [comment]
Thanks, @dhvines . Will follow up as you suggested above. 
Can you please address the question about creating a malicious scenario? Wondering if you had any specific suggestions there. Thank you.

**Comment by dhvines** [comment]
I need a little more context. What do you mean by a malicious scenario?

Sent from my iPhone

On Apr 28, 2020, at 3:53 PM, gitdhar <notifications@github.com> wrote:

﻿

Thanks, @dhvines<https://github.com/dhvines> . Will follow up as you suggested above.
Can you please address the question about creating a malicious scenario? Wondering if you had any specific suggestions there. Thank you.

**Comment by gitdhar** [comment]
@dhvines, example malicious scenarios would be 
a) a user accessing data that he/she is not supposed to
b) a user extracting huge amount of data
We are interested in simulating this and make sure we detect and block such a malicious scenario. Given you are most familiar with the app, your thoughts about this would be most helpful.

**Comment by dhvines** [comment]
That would require some coding changes as there are no authentication and authorization checks in any of the methods. Would you be interesting in adding that new feature?

Sent from my iPhone

On Apr 29, 2020, at 12:22 PM, gitdhar <notifications@github.com> wrote:

﻿

@dhvines<https://github.com/dhvines>, example malicious scenarios would be
a) a user accessing data that he/she is not supposed to
b) a user extracting huge amount of data
We are interested in simulating this and make sure we detect and block such a malicious scenario. Given you are most familiar with the app, your thoughts about this would be most helpful.

**Comment by gitdhar** [comment]
Yes; we would like to work on that. Please give us a few pointers to start with.

**Comment by dhvines** [comment]
I would look into spring security - roles and privileges. Here's a link <https://www.baeldung.com/role-and-privilege-for-spring-security-registration> to get you started
[https://www.baeldung.com/wp-content/uploads/2017/08/Spring-Security-1.jpg]<https://www.baeldung.com/role-and-privilege-for-spring-security-registration>
Spring Security - Roles and Privileges | Baeldung<https://www.baeldung.com/role-and-privilege-for-spring-security-registration>
How to map Roles and Privileges for a Spring Security application: the setup, the authentication and the registration process.
www.baeldung.com
Don

________________________________
